### ISSUE TITLE
Subscription information not updated on profile immediately, starting v1.213.5

### ISSUE THREAD

**Issue by data-sync-user** [mozilla/fxa#11269]
TL;DR - Guardian statsd events seem to indicate a regression in profile cache-ing starting v1.213.5

cc [~accountid:6113f9a38c15ca006fe13cee] who spotted this.

- - -
Guardian is an oauth RP customer of FxA that serves the Mozilla VPN clients. It is a private repo hosted here: [https://github.com/mozilla-services/guardian-website](https://github.com/mozilla-services/guardian-website). We have a function FXA.getUserProfile ([https://github.com/mozilla-services/guardian-website/blob/main/server/lib/fxa.ts#L94-L133)](https://github.com/mozilla-services/guardian-website/blob/main/server/lib/fxa.ts#L94-L133))

In semi-pseudocode it does the following:

```
GET oauthUrls (https://accounts.firefox.com/.well-known/openid-configuration)
GET oauthUrls.userinfo_endpoint - https://profile.accounts.firefox.com/v1/profile
fxaProfile = response.body

// TODO: FIXME: Temporary workaround to FxA caching issue.
// See https://github.com/mozilla/fxa/issues/2463 
if (!fxaProfile.subscriptions) {
  GET https://profile.accounts.firefox.com/v1/subscriptions
  if (this returns subscriptions) {
    log a statsd event
    update the fxaProfile with the subscriptions info
  }
}
return fxaProfile
```

The TODO was written 2 years ago and the FxA issue referenced was closed Oct 2019. However, due to tech debt accumulation we had not got round to doing this clean up.

[~accountid:6113f9a38c15ca006fe13cee] started looking into cleaning this up and noticed the statsd metric and thought to check it. Here is [the graph](https://earthangel-b40313e5.influxcloud.net/d/nc6RwIt7z/all-statsd-events-wip?orgId=1&from=now-1y&to=now&viewPanel=5).

!Screenshot from 2021-12-07 17-20-13.png|width=200,height=183!

We had no hits on this metric until August 25, which coincides with v1.213.5 deploy.

So, it appears that the original issue has been re-birthed and fxaProfile doesn't always get updated with subscriptions immediately.

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/FXA-4297)
┆Attachments: <a href="https://mozilla-hub.atlassian.net/rest/api/2/attachment/content/18444">Screenshot from 2021-12-07 17-20-13.png</a> | <a href="https://mozilla-hub.atlassian.net/rest/api/2/attachment/content/22776">Screenshot from 2022-04-12 15-13-38.png</a> | <a href="https://mozilla-hub.atlassian.net/rest/api/2/attachment/content/22775">Screenshot from 2022-04-12 15-14-56.png</a>

**Comment by data-sync-user** [comment]
➤ Benjamin Bangert commented:

Can you clarify on the actual bug? Is the issue that someone is buying a subscription and the profile endpoint doesn’t reflect that immediately? Or is this something about renewals and the profile endpoint claims there are no valid subscriptions?

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

I understand what you’re asking and can’t give you as much information as you’d like. What I can say is:

We are making a call to /v1/profile we call the response body fxaProfile. if (!fxaProfile.subscriptions) then we make a call to /v1/subscriptions. If that second call has a body on it with subscriptions info then we use this extra info to update our record of the profile and we record the event subscriptionsOverride. You can see that this event started firing at a noticeable rate at around the time of the conversion funnel update and then even higher in the last couple of months.

I am assuming that the issue is that “someone is buying a subscription and the profile endpoint doesn’t reflect that immediately.” But it could be something else. I’m happy to add extra logging to help track this down so we can see if it’s a specific case or subscription type that causes this.

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

The original issue you fixed that was the reason we had this workaround in our code in the first place was https://github.com/mozilla/fxa/issues/2463 ( https://github.com/mozilla/fxa/issues/2463|smart-link ) 

You subsequently fixed that force cache and we should have cleaned out our workaround at the time. But we didn’t and now it appears that the work around is needed again.

**Comment by data-sync-user** [comment]
➤ Bianca Danforth commented:

Sarah Bird Thanks for the clarification. I think I understand what you’re saying – so IIUC, you are observing that the workaround is needed more and more, and you’re wondering why? Is this changed behavior blocking anything critical from your end?

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

Bianca Danforth No, it is not blocking anything critical on our end. Our next release is in early-to-mid jan. Just let us know by, say, Jan 14th, if you want any logging added.

**Comment by data-sync-user** [comment]
➤ Bianca Danforth commented:

Thanks Sarah.

I only see dependency bumps for the profile server changelogs ( https://github.com/mozilla/fxa/blob/main/packages/fxa-profile-server/CHANGELOG.md#12135 ) around this time (1.213 - 1.215), though we do have some possibly related changes in the auth server changelogs.

Looking at the auth server changelogs, it looks like there were a few patches that touch JWT tokens, new subscription create endpoints for the passwordless flow, etc. This is right at the end of finishing development for the Conversion Funnel Optimization epic ( https://mozilla-hub.atlassian.net/browse/FXA-3486 ).

From reviewing prior discussions, it seems like we’d want to call profileClient.deleteCache(uid) which we do for stripeUpdate in the Capability Service, but it looks like in some other places (notably: customerChanged ), we make some additional calls. Is profileClient.deleteCache(uid) sufficient?

Benjamin Bangert It looks like you have been involved in prior discussions regarding profile caching issues in the past. Would you have any thoughts – particularly around what might have changed around that time?

**Comment by data-sync-user** [comment]
➤ Benjamin Bangert commented:

Sarah Bird Relay, MDN, and Hubs have added FxA. None of them are using this workaround, and none of them have reported issues with customers not being able to access the product they purchased, nor in their testing of the flows.

There’s a few things that would help for this to move forward. Can you log what resulting subscriptions found by this workaround? It’s worth noting that the check its doing to increment this counter is:

if (resp.statusCode === 200 && resp.body.subscriptions) {The problem is, the subscriptions endpoint always returns this property, its just an empty list if there’s no subscriptions. While the profile endpoint doesn’t include the property at all.

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

Benjamin Bangert thank you for spotting our bug and spelling out the difference between /profile and /subscriptions APIs for us. We’ll re-instrument and update in ~1 month to see if we’re still seeing the problem.

cc Jeremy Swinarton

**Comment by data-sync-user** [comment]
➤ Benjamin Bangert commented:

Jeremy Swinarton Any update on this? Can we close it?

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

Hi, sorry for the delay. TL;DR Yes

For posterity, here is the graph since we implemented the bugfix you suggested Benjamin Bangert. 

!Screenshot from 2022-04-12 15-13-38.png|width=1668,height=688!

Worth nothing that that line isn’t literally zero. Zooming in to the last 30 days, you can see we still hit it a couple of times a day

!Screenshot from 2022-04-12 15-14-56.png|width=1644,height=676!

So this means that, a couple of times per day, the main profile call returned no subscriptions but then a manual call to the /subscriptions endpoint did then return a subscription.

My guess would be that calling to the profile endpoint again after a short delay would likely fix the issue but we haven’t tried that.

So yes, you can close, thank you for helping us with this.

**Comment by data-sync-user** [comment]
➤ Benjamin Bangert commented:

Thanks!
